### PR TITLE
[2단계 - 장바구니] 비녀(강윤호) 미션 제출합니다.

### PR DESCRIPTION
**Pull request by KangYunHo1221**
[데모사이트](https://kangyunho1221.github.io/react-shopping-cart/)

안녕하세요 오스틴! 반가워요

잘 지내셨는지 모르겠습니다 :)

저는 개인적으로 이번 미션이 시간을 많이 잡아먹어서 가장 밤을 많이 샌 한 주가 아니었나 싶네요 ㅋㅋ

step1과 거의 동일하게 진행하였고, 장바구니에 있는 상품의 수량을 확인할 수 있는 page만 추가하였습니다.

------------------------------

### 😵진행하면서 생각했던 사항

1.
쿠팡같은 서비스를 보면 나갔다 들어와도 구매가 유지되는거 같아서 db의 cartlist에 willPurchase로 넣어줬습니다.
상품의 수량을 변경하거나, check버튼을 누를때마다 통신을 해야해서 안좋을 수도 있다고 생각하지만,
페이지가 정보를 담고 있고, 한번에 업데이트 해주는 방식이면 single source of truth 가 아니라고 생각해서,
매번 통신을 하더라도 일관되게 유지되게 노력했습니다.

2. 단위테스트를 reducer에 대한 간단한 것들을 위주로 작성했습니다.
사실 시간이 많지 않아서 reducer의 모든 action과 edge case를 cover하지는 못한 것 같습니다만..
혹시 reduce이외에 것을 test 한다면 어떤 것을 하는게 좋을까요?
오스틴 님은 꼭 해야한다면 어떤 것을 테스트 하시는 편인지 궁금합니다.

------------------------------
## 🤔 Issues

1. ~~원래 서버랑 통신할때는 장바구니 수량이 변경될때마다 잘 re-render 해줘서 수량확인이 됐었는데, msw를 사용하니까 작동하지 않는 버그가 있네요.
최대한 빨리 수정하도록 하겠습니다.~~
수정하였습니다

2. msw를 이용한 devlopment에서는 정상 작동하는데 전체선택을 사용하면 json-server가 터져버리는 이슈가 있습니다.
전체선택기능이 전체 list 를 순회하면서 check와  uncheck를 실행시켜주는 방식이라 갑자기 많은양의 통신이 발생하면
배포 서버가 감당이 안되는 것 같습니다 :(

3. 장바구니 페이지에서 수량 변경시 깜빡이듯이 리렌더링 되는 문제가 있습니다.
버튼 이벤트마다 통신을 하기 때문에 최적화가 어려울 것 같아서
쿠팡처럼 이전 상태 위에 loading Spinner를 보여주고 loading이 완료 된 후 새로운 페이지를 렌더해주는 방식으로 처리하도록 하려고 합니다,.. 약간은 꼼수같다는 생각이 들지만 현재는 더 좋은 방법이 떠오르지 않네요

---------------------------------

### 😹

이번 미션이 개인적으로는 역대급으로 버그가 많았던(msw관련 설정 이슈를 해결하느라) 미션이라서 리팩토링을 거의 하지 못한 것 같네요.
보통 짜고 나서 찬찬히 돌아보면서 리팩토링 하는 편인데, 짤 때부터 깔끔하게 짜는 습관을 들여야 겠다고 느꼈습니다.

### CODE REVIEW

**Review by austinpark420 — CHANGES_REQUESTED** [review]
안녕하세요. 비녀 이번 미션은 힘들게 작업하셨나보군요 😢 
코드 관련해서 수정하면 좋을 내용 코멘트 남겨놨습니다. 확인 부탁드릴게요!!

**Comment by austinpark420** [comment]
App.tsx, header.tsx 에서 라우팅을 할때는 상수를 사용하시는게 좋아요! 

```html
// header.tsx
    <StyledRoot>
      <div>
        <Link to='/main/1'>
          <StyledLogo>
            <CartIcon fill='white' />
            <StyledBrandName>RoyStin Shop</StyledBrandName>
          </StyledLogo>
        </Link>
        <StyledNav>
          <button onClick={() => navigate('/cart')}>장바구니</button>
          <button>주문목록</button>
        </StyledNav>
      </div>
    </StyledRoot>

// app.tsx
  <BrowserRouter basename={process.env.PUBLIC_URL}>
      <StyledRoot>
        <Header />
        <StyledMain>
          <Routes>
            <Route path='/' element={<Navigate replace to='/main/1' />} />
            <Route path='/main/:id' element={<ItemListPage />} />
            <Route path='/item_detail/:id' element={<ItemDetail />} />
            <Route path='/cart' element={<CartPage />} />
            <Route path='*' element={<NotFound />} />
          </Routes>
        </StyledMain>
      </StyledRoot>
      {isSnackbarOpen && <Snackbar />}
    </BrowserRouter>
```

**Comment by austinpark420** [comment]
> 쿠팡같은 서비스를 보면 나갔다 들어와도 구매가 유지되는거 같아서 db의 cartlist에 willPurchase로 넣어줬습니다.
상품의 수량을 변경하거나, check버튼을 누를때마다 통신을 해야해서 안좋을 수도 있다고 생각하지만,
페이지가 정보를 담고 있고, 한번에 업데이트 해주는 방식이면 single source of truth 가 아니라고 생각해서,
매번 통신을 하더라도 일관되게 유지되게 노력했습니다.

쿠팡도 수량변경할 때마다 api 통신을 하고 있네요! 안좋을 수 있다고 생각하는 부분은 너무 많은 api 호출 때문이라고 생각하시는건가요?!

>단위테스트를 reducer에 대한 간단한 것들을 위주로 작성했습니다.
사실 시간이 많지 않아서 reducer의 모든 action과 edge case를 cover하지는 못한 것 같습니다만..
혹시 reduce이외에 것을 test 한다면 어떤 것을 하는게 좋을까요?
오스틴 님은 꼭 해야한다면 어떤 것을 테스트 하시는 편인지 궁금합니다.

테스트와 크게 다른 부분은 없습니다. 서비스 플로우에 맞게 작성해주시면 됩니다. 딱히 어떤걸 해야한다는 것 없는  것 같습니다!

**Comment by austinpark420** [comment]
사용하지 않는 컴포넌트나 함수의 파라미터 등 제거를 안하신 경우가 좀 있는 것 같습니다! 다음 미션때는 바쁘시더라도 코드보시면서 정리해주시는게 좋을 것 같습니다!!

File: src/App.tsx [hunk 1]
```diff
@@ -6,7 +6,7 @@ import styled from 'styled-components';
 import Header from 'components/common/Header';
 import Snackbar from 'components/common/Snackbar';
 import { useAppSelector } from 'hooks/useAppSelector';
-import Cart from 'pages/CartPage';
+import CartPage from 'pages/CartPage';
 import useSnackBar from 'hooks/useSnackBar';
```
[the remark this answers is not in the document]
Comment (KangYunHo1221) [reply src/App.tsx:10]: SnackBar의 구조가 좀 복잡한데요,

redux의 snackbar reducer 에서 isOpen상태를 관리하고
Portal 에서 createportal할지 null 할지를 isOpen에 따라서 결정
useSnackbar 훅을 통해서 dispatch를 관리
snackbar 컴포넌트는 어떤걸 보여줄지 contents만 관리

이런 구조와 역할을 가지고 있습니다.

다음과 같은 고민을 통해서 만들게 됐습니다.

1. modal, snackbar, toast는 본질적으로 '메인화면 위에 새 창을 띄우는 점'이 같다고 생각
2. 새 창을 띄우는 것을 portal로 만들어 주자
3. portal의 상태를 전역으로 관리하면 
 - 어떤 component에서든 쉽게 접근해서 portal을 띄울 수 있으니 범용성이 좋을 것이다.
 - 여러 곳에서 사용했을 시 snackbar 마다 상태가 여러개가 생성돼서 관리하기 어려울 것이다. 전역으로 하나의 상태만 두도록 하자.

4. Snackbar 컴포넌트와 useSnackbar hook을 분리해서 사용하는 이유

```javascript
// 아무 컴포넌트나 사용하고 싶은 곳
const {openSnackbar} = useSnackbar();
const onClick = () =>{openSnackbar('이 타입의 모달창을 띄워줘')}
```

이렇게 쉽게 관리해서 쓰고 싶었습니다.

[the remark this answers is not in the document]
Comment (KangYunHo1221) [reply src/App.tsx:10]: 네이밍이 좀 헷갈리는 것 같아서 createPortal 할 위치를 root-modal 로 변경하였고,
app에서도 snackbar나 modal창이 여러개 생기면 문제가 생길 것 같아서 이름을 변경하였습니다.
snackBar도 Modal이 더 통일감 있는 이름인것 같아서 변경했습니다.

File: src/redux/reducers/cartListReducer.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import { CartListAction, CartListActionType } from 'redux/actions/cartList';
```
[the remark this answers is not in the document]
Comment (KangYunHo1221) [reply src/redux/reducers/cartListReducer.ts:1]: 그거 관련된 설정도 있나요...? 😭
못 찾아서 전부 띄어쓰기 해주도록 수동으로 변경했습니다..

File: src/pages/CartPage.tsx [hunk 3]
```diff
@@ -1,11 +1,93 @@
...
+      .map(item => {
+        return `id=${item.id}`;
+      })
+      .join('&');
+
+  const {
+    data: cartDetail,
+    loading: loadingGetCartDetail,
+    error: errorGetCartDetail,
+  }: { data: Item[]; loading: boolean; error: string } = useFetch(cartListString);
```
[the remark this answers is not in the document]
Comment (KangYunHo1221) [reply src/pages/CartPage.tsx:37]:
```javascript

"cartList": [
    {
      "id": 3,
      "quantity": 1,
      "willPurchase": true
    }

vs

"cartList": [
    {
      "id": 3,
      "thumbnailUrl' : 'html://....'
      "title" : 'product name'
      "quantity": 1,
      "willPurchase": true
    }

```
실제 서비스들도 많이 봤는데, 쿠팡은 id랑 quantity만 가지고 무신사는 id랑 detail들도 가지고 하는 식으로

서비스마다 표준화 된 건 없는 것 같았습니다.


그래서 위의 두 가지 구조중에 어떤 방식을 선택할지 고민을 많이 했었습니다.

페어와 나눈 결론은 여기서도 sigle source of truth를 최대한 구현하려고 노력하자 였는데요,

data를 관리할때 thumbnail url이 itemList에도 있고, cartList에도 있는 것보다는

cartList에서는 id와 quantity등 최소한의(겹치지 않는) 정보만 제공하고 필요시에 항상 조회를 하는게 좋지 않을까 생각해서 지금과 같은 형태가 됐습니다.

File: src/components/Cart/CartList.tsx [hunk 4]
```diff
@@ -1,29 +1,219 @@
...
+    });
+  };
+
+  const toggleChecked = (targetId: number) => {
+    toggleCartItemWillPurchase(targetId);
+  };
+
+  const modifyQuantity = (targetId: number, type: 'up' | 'down' | 'alter', quantity: number) => {
+    updateCartItemQuantity(targetId, type, 1);
+  };
```
[the remark this answers is not in the document]
Comment (KangYunHo1221) [reply src/components/Cart/CartList.tsx:64]: 원래 구현하려고 하던 기능중에 controller의 숫자를 변경하면 modify를 바로 적용하도록 할 계획으로 넣어줬었습니다.

현재는 필요하지 않으니 삭제하도록 하겠습니다.

File: src/components/Cart/CartList.tsx [hunk 5]
```diff
@@ -1,29 +1,219 @@
...
+              <ItemName>{detail.title}</ItemName>
+              <StyledRight>
+                <TrashCan id={cartItem.id} onClick={() => deleteItem(id)} />
+                <Controller
+                  id={cartItem.id}
+                  quantity={cartItem.quantity}
+                  modifyQuantity={modifyQuantity}
+                ></Controller>
+                <TotalPrice>
+                  {totalPrice.toString().replace(/\B(?<!\.\d*)(?=(\d{3})+(?!\d))/g, ',')} 원
```
[the remark this answers is not in the document]
Comment (KangYunHo1221) [reply src/components/Cart/CartList.tsx:108]: 좋은 것 같습니다. 한군데서만 사용한다고 생각했는데 쓰일 곳이 꽤 많은 것 같아서 utils/index.js 로 분리하였고,

itemList/ItemContainer에서도 사용 가능 한 것 같아서 금액이 사용되는 곳에 전부 적용해 주었습니다!!

**Comment by KangYunHo1221** [comment]
새로 배우는 것들이 너무 많다보니 적용하는데만 급급했던 것 같습니다.

기본적인 것들을 많이 못지킨 것 같아서 죄송한 마음이 드네요.ㅠㅠ 😭

살짝 늦더라도 조금 더 신경써서 해야 했는데,, 다음부턴 더 신경쓰도록 하겠습니다. 

- 라우팅은 라우터 컴포넌트를 분리해서 묶어주고, 해당 폴더에 경로를 상수화 해서 넣어주었습니다.
상수화 해서 실수도 줄이고, 라우터도 분리하면 app의 역할이 줄어들 것 같아서 분리했습니다!
 
- 사용하지 않는 컴포넌트나 함수의 파라미터 등 제거를 안하신 경우가 좀 있는 것 같습니다!
이건 제 불찰이 너무너무 크네요... 다음부터는 많이 신경써서 제출하도록 하겠습니다.

- 쿠팡도 수량변경할 때마다 api 통신을 하고 있네요! 안좋을 수 있다고 생각하는 부분은 너무 많은 api 호출 때문이라고 생각하시는건가요?!
네 맞습니다! 통신을 어느정도까지 해야하는지 아직 감이 잘 없는 것 같아요.



수정사항 반영했고, 코멘트 남겼습니다!!

리뷰 남겨주셔서 감사합니다!

추가로 봤으면 좋겠는 사항이 있으면 가감없이 패주셔도 됩니다!!!

File: src/pages/CartPage.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (austinpark420) [reply src/pages/CartPage.tsx:37]: sigle source of truth!! 상세하게 설명주셔서 감사합니다. 그런 의도였군요 👍

**Comment by austinpark420** [comment]
> 쿠팡도 수량변경할 때마다 api 통신을 하고 있네요! 안좋을 수 있다고 생각하는 부분은 너무 많은 api 호출 때문이라고 생각하시는건가요?!
> ->네 맞습니다! 통신을 어느정도까지 해야하는지 아직 감이 잘 없는 것 같아요.

회사 정책에 따라 다르겠지만 api 호출이 많더라고 해야하는 것 같습니다 ㅎㅎ

File: src/App.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (austinpark420) [reply src/App.tsx]: > useSnackbar 훅을 통해서 dispatch를 관리
> snackbar 컴포넌트는 어떤걸 보여줄지 contents만 관리

useSnackbar에서 보여주는 컨테츠 관리를 할 수 있으면 snackbar가 필요없을 것 같은데 어떻게 생각하세요?

File: src/redux/reducers/cartListReducer.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (austinpark420) [reply src/redux/reducers/cartListReducer.ts]: 앗 이부분은 제가 착각했네요.. 어느정도는 수동으로 맞춰줘야하는군요 🙏

**Review by austinpark420 — APPROVED** [review]
안녕하세요 비녀!

깔끔하게 코드 수정해주셨네요 👍 
이번 미션은 여기서 머지를 해도 좋을 것 같습니다.
간단하게 코멘트 남겨놨는데 확인하시고 추가로 궁금한 사항 있으면 말씀부탁드려요 😄
